### PR TITLE
chore: move address storage to backend and return last unused address

### PR DESCRIPTION
**Pull request by rolznz**
Fixes https://github.com/getAlby/nostr-wallet-connect-next/issues/384

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -375,19 +376,62 @@ func (api *api) CloseChannel(ctx context.Context, peerId, channelId string, forc
 	})
 }
 
-func (api *api) GetNewOnchainAddress(ctx context.Context) (*NewOnchainAddressResponse, error) {
+func (api *api) GetNewOnchainAddress(ctx context.Context) (*OnchainAddressResponse, error) {
```
Comment (rolznz) [inline api/api.go:379]: Update: I will change this to a string to be consistent

**Comment by bumi** [comment]
is it worth adding all this code?

**Comment by rolznz** [comment]
> is it worth adding all this code?

In my opinion, yes:
- No longer reliant on local storage
- Automatically choose a new address when the old one was used
- Remove a lot of duplication of onchain address management in the frontend

File: api/api.go [hunk 2]
```diff
@@ -375,17 +376,58 @@ func (api *api) CloseChannel(ctx context.Context, peerId, channelId string, forc
...
+		api.logger.WithError(err).Error("Failed to get current address from config")
+		return "", err
+	}
+
+	if currentAddress != "" {
+		// check if address has any transactions
+		response, err := api.RequestEsploraApi("/address/" + currentAddress + "/txs")
+		if err != nil {
+			api.logger.WithError(err).Error("Failed to get current address transactions")
+			return "", err
```
Comment (bumi) [inline api/api.go:409]: can we create a new one if the call fails? 
to not prevent the execution of the rest of the code if esplora is down.

File: api/api.go [hunk 3]
```diff
@@ -375,17 +376,58 @@ func (api *api) CloseChannel(ctx context.Context, peerId, channelId string, forc
...
+		}
+
+		api.logger.WithField("response", response).Info("Got address response")
+		transactions, ok := response.([]interface{})
+		if !ok {
+			err = fmt.Errorf("Failed to cast esplora address txs response: %v", response)
+			return "", err
```
Comment (bumi) [inline api/api.go:416]: same here, does it make sense to return a new one if espora does not behave nicely?

File: api/models.go [hunk 4]
```diff
@@ -29,7 +29,8 @@ type API interface {
 	DisconnectPeer(ctx context.Context, peerId string) error
 	OpenChannel(ctx context.Context, openChannelRequest *OpenChannelRequest) (*OpenChannelResponse, error)
 	CloseChannel(ctx context.Context, peerId, channelId string, force bool) (*CloseChannelResponse, error)
-	GetNewOnchainAddress(ctx context.Context) (*NewOnchainAddressResponse, error)
+	GetNewOnchainAddress(ctx context.Context) (string, error)
+	GetLastUnusedOnchainAddress(ctx context.Context) (string, error)
```
Comment (bumi) [inline api/models.go:33]: maybe GetUnusedOnchainAddress? 
what means "last"?

**Review by bumi — APPROVED** [review]

File: api/api.go (hunk 2 above)
Comment (rolznz) [reply api/api.go]: I updated both places, let me know if that's ok for you. Otherwise, I'll change them to get a new address instead

**Pull request merged**